### PR TITLE
[US2422815] Protobuf Enum Support

### PR DESCRIPTION
**Pull request by WesleyTangNationalInstruments**
- [x] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/measurement-services-python/blob/main/CONTRIBUTING.md).

### What does this Pull Request accomplish?

- Adds support for using protobuf enums with MeasurementLink.
- Updated sample_measurement to include an example of protobuf enums

### Why should this Pull Request be merged?

[AB#2422815](https://ni.visualstudio.com/DevCentral/_backlogs/backlog/InstrumentStudio%20Platform/Features/?workitem=2422815)

### What testing has been done?

- Manually verified sample_measurement using protobuf enums runs in InstrumentStudio.
- Manually verified sample_measurement using protobuf enums are linked to the correct controls and indicators in MeasurementUIEditor.
- Verified tests pass.

image
image

### CODE REVIEW

File: examples/sample_measurement/assets/color_pb2.pyi
Comment (bkeryan) [inline examples/sample_measurement/assets/color_pb2.pyi:1]: Did you generate this by running protoc manually? Please update `scripts/generate_grpc_stubs.py`

File: ni_measurementlink_service/_internal/parameter/serializer.py [hunk 1]
```diff
@@ -224,11 +227,18 @@ def _deserialize_enum_parameters(
         parameter_metadata = parameter_metadata_dict[id]
         if get_enum_values_annotation(parameter_metadata):
             enum_type = _get_enum_type(parameter_metadata)
+            is_protobuf_enum = enum_type is int
```
Comment (bkeryan) [inline ni_measurementlink_service/_internal/parameter/serializer.py:230]: Please test `IntEnum` as well. An `IntEnum` is an instance of both `Enum` and `int`.

File: ni_measurementlink_service/_internal/parameter/serializer.py [hunk 2]
```diff
@@ -75,15 +76,17 @@ def serialize_parameters(
...
             == TypeSpecialization.Enum.value
         ):
             if parameter_metadata.repeated:
-                parameter = [x.value for x in parameter]
+                if len(parameter) > 0 and isinstance(parameter[0], Enum):
+                    parameter = [x.value for x in parameter]
             else:
-                parameter = parameter.value
+                if isinstance(parameter, Enum):
+                    parameter = parameter.value
```
Comment (bkeryan) [inline ni_measurementlink_service/_internal/parameter/serializer.py:89]: It would be more readable to call a `_get_enum_value` function that does the `isinstance` check because then you don't have to think about what it means to skip assigning to `parameter`.

File: ni_measurementlink_service/measurement/service.py [hunk 3]
```diff
@@ -7,9 +7,10 @@
 from os import path
 from pathlib import Path
 from threading import Lock
-from typing import Any, Callable, Dict, List, Optional, Type, TypeVar
+from typing import Any, Callable, Dict, List, Optional, Type, TypeVar, Union
 
 import grpc
+from google.protobuf.internal.enum_type_wrapper import EnumTypeWrapper
```
Comment (bkeryan) [inline ni_measurementlink_service/measurement/service.py:13]: This is internal, so I'm not sure we're allowed to reference it.

If we must import an internal type, we can reduce the risk of breakage by doing the import only when type checking, not at run time.

```
from __future__ import annotations
...
import typing
...

if typing.TYPE_CHECKING:
   from google.protobuf.internal.enum_type_wrapper import EnumTypeWrapper
```

File: ni_measurementlink_service/measurement/service.py [hunk 4]
```diff
@@ -370,12 +375,20 @@ def _make_annotations_dict(
 
         return annotations
 
-    def _enum_to_annotations_value(self, enum_type: Type[Enum]) -> str:
-        if not any(member.value == 0 for member in enum_type):
-            raise ValueError("The enum does not have a value for 0.")
+    def _enum_to_annotations_value(
+        self, enum_type: Union[Type[Enum], Type[EnumTypeWrapper]]
```
Comment (bkeryan) [inline ni_measurementlink_service/measurement/service.py:379]: Consider defining a type alias instead of duplicating this type.

File: ni_measurementlink_service/measurement/service.py [hunk 5]
```diff
@@ -370,12 +375,20 @@ def _make_annotations_dict(
...
+    ) -> str:
         enum_values = {}
-        for member in enum_type:
-            enum_values[member.name] = member.value
+        if isinstance(enum_type, EnumTypeWrapper):
```
Comment (bkeryan) [inline ni_measurementlink_service/measurement/service.py:382]: Should this be `issubclass`? Or is an instance of `EnumTypeWrapper` itself a type?

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/sample_measurement/assets/color_pb2.pyi
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/assets/color_pb2.pyi:1]: Updated `scripts/generate_grpc_stubs.py` to automatically generate these files.

File: ni_measurementlink_service/_internal/parameter/serializer.py (hunk 1 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/_internal/parameter/serializer.py:225]: Added tests for IntEnum.

File: ni_measurementlink_service/_internal/parameter/serializer.py (hunk 2 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/_internal/parameter/serializer.py]: Added `_get_enum_value` helper as suggested.

File: ni_measurementlink_service/measurement/service.py (hunk 4 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py]: Added `SupportedEnumType` alias.

File: ni_measurementlink_service/measurement/service.py (hunk 5 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py]: The enum_type that gets passed in for protobuf enums appears to be an instance of EnumTypeWrapper in addition to being a type itself which is why we used `isinstance` here. I'm not entirely sure why the protobuf enum types are instances of EnumTypeWrapper rather than just being a subtype of it, but that seems to be the way it behaves. 

In our decorator, the intent was for users to pass in that type directly.

`@sample_measurement_service.configuration("Protobuf Enum In", nims.DataType.Enum, ProtobufColor.BLACK, enum_type=ProtobufColor)`

If instead we wanted to use `issubclass(enum_type, EnumTypeWrapper)` here, we would need the users to pass in `enum_type=type(ProtobufColor))` instead. 

I updated this code to check `issubclass(type(enum_type), EnumTypeWrapper)` and added a comment to make it clearer, but note that `type(enum_type)` is required since `issubclass` throws an error if the parameter is an instance.

Comment (bkeryan) [reply ni_measurementlink_service/measurement/service.py:392]: I'm not sure whether the enum type wrapper instance is a type or not.

The `.pyi` file says it subclasses `builtins.type`:
https://github.com/ni/nidaqmx-python/blob/master/generated/nidaqmx/_stubs/session_pb2.pyi

But at run time, `isinstance` says it's not a type.
```
>>> from nidaqmx._stubs import nidaqmx_pb2
>>> session_pb2.SessionInitializationBehavior
<google.protobuf.internal.enum_type_wrapper.EnumTypeWrapper object at 0x000001D87531E8E0>
>>> session_pb2.SessionInitializationBehavior.keys()
['SESSION_INITIALIZATION_BEHAVIOR_UNSPECIFIED', 'SESSION_INITIALIZATION_BEHAVIOR_INITIALIZE_NEW', 'SESSION_INITIALIZATION_BEHAVIOR_ATTACH_TO_EXISTING']
>>> type(session_pb2.SessionInitializationBehavior)
<class 'google.protobuf.internal.enum_type_wrapper.EnumTypeWrapper'>
>>> isinstance(session_pb2.SessionInitializationBehavior, type)
False
```

I wonder if `SupportedEnumType = Union[Type[Enum], EnumTypeWrapper]` would also work.

Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py:392]: Yeah, I saw this too and I'm not entirely sure why it behaves this way. Despite the `type()` of it being a class, issubclass does not believe it to be so. 

image

Note that `SupportedEnumType = Union[Type[Enum], EnumTypeWrapper]` does also work. Let me know if you prefer this to `Union[Type[Enum], Type[EnumTypeWrapper]]` and I can make the change, but I've left it as-is for now since I'm not sure it really makes it any clearer.

File: examples/sample_measurement/measurement.py [hunk 6]
```diff
@@ -4,6 +4,7 @@
 from enum import Enum
 
 import click
+from examples.sample_measurement.assets.color_pb2 import ProtobufColor
```
[the remark this answers is not in the document]
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/measurement.py:7]: Hmm, I'm not sure how to best resolve this without trying to separate the acceptance tests from using sample_measurement service. 

I thought we could include the assets in the examples.zip and then reference `assets.color_pb2` but I don't think that works if we don't make assets an module because the tests break. Perhaps we can create a module that contains just `colors.proto` and associated files that sample_measurement and tests can use, but that seems a bit excessive. We could remove the protobuf enum from sample_measurement and implement test coverage at a lower level as an alternative.

Note that importing `examples.sample_measurement.assets.color_pb2` does work when running the example directly and with running tests on my machine, but I suppose that's only because I'm running it out of the repo where the assets files can be found in the examples directory.

File: ni_measurementlink_service/measurement/service.py (hunk 3 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py:14]: Let me know if I'm misinterpreting this situation, but only importing during type checking doesn't seem to work as we use `EnumTypeWrapper` for more than just `SupportedEnumType`. We reference `EnumTypeWrapper` as part of generating the annotations for the enum (in `MeasurementService._enum_to_annotations_value`) when we resolve the configuration and output decorators. I tried to find a public alternative to `EnumTypeWrapper` that we could use instead to effectively make the same check but couldn't find anything that panned out. I've left this as-is for now.

File: examples/sample_measurement/measurement.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bkeryan) [reply examples/sample_measurement/measurement.py:7]: Let's do a conditional import for now and consider separating the examples from the tests later.

```
try:
    from .assets import color_pb2
except ImportError:
    from examples.sample_measurement.assets import color_pb2
```

File: ni_measurementlink_service/measurement/service.py [hunk 7]
```diff
@@ -377,12 +384,21 @@ def _make_annotations_dict(
...
-        for member in enum_type:
-            enum_values[member.name] = member.value
+        # Note that the type of protobuf enums are an instance of EnumTypeWrapper
+        # in addition to being a type itself. Additionally, issubclass excludes
+        # instances as valid parameters so we use isinstance here.
+        if isinstance(enum_type, EnumTypeWrapper):
+            if not any(value.number == 0 for value in enum_type.DESCRIPTOR.values):
+                raise ValueError("The enum does not have a value for 0.")
+            for value in enum_type.DESCRIPTOR.values:
+                enum_values[value.name] = value.number
```
Comment (bkeryan) [inline ni_measurementlink_service/measurement/service.py:396]: `EnumTypeWrapper` provides `dict`-style methods that are more concise than accessing the `EnumDescriptor` directly.

```
            if 0 not in enum_type.values():
                raise ValueError("The enum does not have a value for 0.")
            for name, value in enum_type.items():
                enum_values[name] = value
```

File: ni_measurementlink_service/measurement/service.py (hunk 5 above)
Comment (bkeryan) [reply ni_measurementlink_service/measurement/service.py:392]: Yes, I prefer treating it as an instance of `EnumTypeWrapper` because that is true for both run time and type checking.

Also, please remove "in addition to being a type itself" from the comment because this is only true for type checking, not run time.

If you needed to change the order of the if-statement cases, I think `if isinstance(enum_type, type) and issubclass(enum_type, Enum)` would prevent it from throwing.

File: tests/unit/test_metadata.py [hunk 8]
```diff
@@ -25,6 +25,15 @@ class DifferentColor(Enum):
     BROWN = 3
 
 
+class Countries(IntEnum):
+    """Countries enum used for testing enum-typed config and output."""
```
Comment (bkeryan) [inline tests/unit/test_metadata.py:29]: Please call out that this tests `IntEnum`

File: examples/sample_measurement/pyproject.toml [hunk 9]
```diff
@@ -14,6 +14,7 @@ mypy = ">=1.0"
 grpc-stubs = "^1.53"
 # Uncomment to use prerelease dependencies.
 # ni-measurementlink-service = {path = "../..", develop = true}
+types-protobuf = "^4.21"
```
Comment (bkeryan) [inline examples/sample_measurement/pyproject.toml:17]: [style] Let's keep the type stub modules together

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/sample_measurement/pyproject.toml (hunk 9 above)
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/pyproject.toml]: Moved above the commented out dependencies as suggested.

File: tests/unit/test_metadata.py (hunk 8 above)
Comment (WesleyTangNationalInstruments) [reply tests/unit/test_metadata.py]: Updated comment as suggested.

File: ni_measurementlink_service/_internal/parameter/metadata.py [hunk 10]
```diff
@@ -161,5 +161,8 @@ def get_enum_values_annotation(parameter_metadata: ParameterMetadata) -> str:
 
 def _is_valid_enum_value(enum_value: object, user_enum: Dict[str, int]) -> bool:
     if not isinstance(enum_value, Enum):
-        return False
+        if isinstance(enum_value, int):
```
[the remark this answers is not in the document]
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/_internal/parameter/metadata.py]: Simplified as suggested.

File: ni_measurementlink_service/measurement/service.py (hunk 7 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py]: Refactored as suggested. Note that I also had to add a cast here to resolve mypy issue complaining about `Union[Type[Enum], EnumTypeWrapper]` not having `values` or `items` because we swapped the `isinstance` check to not reference `EnumTypeWrapper` directly anymore.

File: ni_measurementlink_service/measurement/service.py (hunk 3 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py]: Updated to import EnumTypeWrapper only on type checking and swapped the runtime `isinstance` usage via the EnumDescriptor instead. Thanks for the suggestion here.

File: ni_measurementlink_service/measurement/service.py (hunk 5 above)
Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py]: Switched to `Union[Type[Enum], EnumTypeWrapper]` as suggested and fixed comment. Fixed condition as suggested to fix mypy error as suggested.

File: examples/sample_measurement/measurement.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/measurement.py]: Switched to using a conditional import as suggested.

File: examples/sample_measurement/assets/color.proto
[the remark this answers is not in the document]
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/assets/color.proto:1]: Added `__init__.py` and renamed directory as suggested.

File: examples/sample_measurement/measurement.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/measurement.py]: @bkeryan I'm hitting some mypy issues when it runs on this PR and I'm not sure how best to resolve them. 

These do not appear when I run locally via `poetry run mypy -p examples`. I tried directly specifying `sample_measurement` via `poetry run mypy .\examples\sample_measurement\` (which is how it looks like the PR checks are running this check). I do get a "Source file found twice under different module names", but I get none of the same errors from this PR's example check. I'm not sure mypy that's being run on this PR likes the conditional import workaround or respecting the `TYPE_CHECK` condition to import `EnumTypeWrapper`. 

Doing some investigating on stackoverflow, I suspect that the PR's mypy is configured differently than what I have locally, but I'm unsure fighting with mypy into ignoring these errors is the right approach. I tried making sample_measurement's a module by adding a `__init__.py` for that directory as well and changing the imports to reference differently without much success. Please let me know if you have any ideas on how to get the examples check passing.

[the remark this answers is not in the document]
Comment (bkeryan) [reply examples/sample_measurement/measurement.py]: These problems are why the GitHub "check examples" works the way it does: https://github.com/ni/measurementlink-python/blob/main/.github/workflows/check_examples.yml

You need to run `poetry install` and `poetry run mypy .` separately for each example.

[the remark this answers is not in the document]
Comment (bkeryan) [reply examples/sample_measurement/measurement.py]: https://github.com/ni/measurementlink-python/actions/runs/5557275444/jobs/10150844869?pr=328

```
  measurement.py:11: error: Cannot find implementation or library stub for module named "examples.sample_measurement._stubs"  [import]
  measurement.py:11: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
  measurement.py:11: error: Name "color_pb2" already defined (by an import)  [no-redef]
```
The first error can be avoided by setting `[[tool.mypy.overrides]]` appropriately: https://github.com/ni/measurementlink-python/blob/main/examples/nidmm_measurement/pyproject.toml#L25

But the second error probably needs `# type: ignore`. 

I think I saw the no-redef error here too but never got around to posting a PR for the fix: https://github.com/ni/nidaqmx-python/blob/master/generated/nidaqmx/__init__.py

```
  measurement.py:[50](https://github.com/ni/measurementlink-python/actions/runs/5557275444/jobs/10150844869?pr=328#step:7:59): error: Argument "enum_type" to "configuration" of "MeasurementService" has incompatible type "type[ProtobufColor]"; expected "Optional[type[Enum]]"  [arg-type]
  measurement.py:61: error: Argument "enum_type" to "output" of "MeasurementService" has incompatible type "type[ProtobufColor]"; expected "Optional[type[Enum]]"  [arg-type]
  Found 4 errors in 1 file (checked 4 source files)
```

This is because the example is using the released `ni-measurementlink-service` 1.1.0 which does not include your changes. You need to uncomment the ../.. stuff again.

File: ni_measurementlink_service/measurement/service.py (hunk 7 above)
Comment (bkeryan) [reply ni_measurementlink_service/measurement/service.py]: You can remove the `if TYPE_CHECKING` around the cast by passing the type name as a string :)

image

Typeshed confirms that `cast` has an overload that takes `typ: str`.
https://github.com/python/typeshed/blob/main/stdlib/typing.pyi#L825

Comment (WesleyTangNationalInstruments) [reply ni_measurementlink_service/measurement/service.py]: Swapped to casting by name rather than type as suggested.

File: examples/sample_measurement/measurement.py
Comment (bkeryan) [inline examples/sample_measurement/measurement.py:1]: In isolation, this PR looks good to go. However, https://github.com/ni/measurementlink-python/pull/329 is decoupling the tests from the examples, so I think you should undo your example changes and reapply them to the new `fake_measurement_service` test utility.

**Review by bkeryan — CHANGES_REQUESTED** [review]

**Review by bkeryan — APPROVED** [review]

File: examples/sample_measurement/measurement.py
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/measurement.py:1]: @bkeryan I think having the protobuf enum in the sample_measurement example brings value to the end user that it is worth it to keep it in. We wanted to ensure that we have examples that covered all of the enum use cases anyways so protobuf enum in the examples would have been something we wanted in anyways. I'd be in favor of checking in this PR now and then either me or @vigkre can update `fake_measurement_service` afterwards if you are fine with that.

Comment (bkeryan) [reply examples/sample_measurement/measurement.py:1]: Ok, then I approve of merging it as-is. Should any other reviewers approve? @dixonjoel @synesthetize

File: examples/sample_measurement/pyproject.toml [hunk 11]
```diff
@@ -12,9 +12,16 @@ click = ">=7.1.2, !=8.1.4" # mypy fails with click 8.1.4: https://github.com/pal
 [tool.poetry.group.dev.dependencies]
 mypy = ">=1.0"
 grpc-stubs = "^1.53"
+types-protobuf = "^4.21"
 # Uncomment to use prerelease dependencies.
-# ni-measurementlink-service = {path = "../..", develop = true}
+ni-measurementlink-service = {path = "../..", develop = true}
```
Comment (dixonjoel) [inline examples/sample_measurement/pyproject.toml:17]: This should be commented out when you check it in.

File: examples/sample_measurement/pyproject.toml [hunk 12]
```diff
@@ -12,9 +12,16 @@ click = ">=7.1.2, !=8.1.4" # mypy fails with click 8.1.4: https://github.com/pal
...
 
 [build-system]
 requires = ["poetry-core>=1.2.0"]
 build-backend = "poetry.core.masonry.api"
+
+[[tool.mypy.overrides]]
+module = [
+    "ni-measurementlink-service.*",
+]
+ignore_missing_imports = true
```
Comment (dixonjoel) [inline examples/sample_measurement/pyproject.toml:27]: Why was this necessary?

Comment (dixonjoel) [reply examples/sample_measurement/pyproject.toml:27]: The ignore_missing_imports, that is.

**Review by dixonjoel — APPROVED** [review]

File: examples/sample_measurement/pyproject.toml (hunk 11 above)
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/pyproject.toml]: Fixed as suggested.

File: examples/sample_measurement/pyproject.toml (hunk 12 above)
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/pyproject.toml]: I had copied this from the other examples' `pyproject.toml` and thought it is intended to solve the missing imports mypy issue, but everything still seems to work after I got rid of it so I've removed it.

Comment (bkeryan) [reply examples/sample_measurement/pyproject.toml]: `ni-measurementlink-service` should not be missing.

File: examples/sample_measurement/measurement.py [hunk 13]
```diff
@@ -38,6 +43,12 @@ class Color(Enum):
 @sample_measurement_service.configuration(
     "Enum In", nims.DataType.Enum, Color.BLUE, enum_type=Color
 )
+@sample_measurement_service.configuration(
+    "Protobuf Enum In",
+    nims.DataType.Enum,
+    color_pb2.ProtobufColor.BLACK,
+    enum_type=color_pb2.ProtobufColor,  # type: ignore
```
Comment (bkeryan) [inline examples/sample_measurement/measurement.py:50]: `# type: ignore` should not be necessary here unless you are testing with the previous release of ni-measurementlink-service, in which case protobuf enums don't work at all.

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/sample_measurement/pyproject.toml (hunk 11 above)
Comment (bkeryan) [reply examples/sample_measurement/pyproject.toml]: I previously asked Wesley to uncomment this so the example will use his update code. Otherwise, it uses ni-measurementlink-service 1.1.0, which doesn't support protobuf enums.

File: examples/sample_measurement/pyproject.toml
Comment (bkeryan) [inline examples/sample_measurement/pyproject.toml:1]: Congrats, this example now requires version 1.2

```
ni-measurementlink-service = {version = "^1.2.0-dev0", allow-prereleases = true}
...
ni-measurementlink-service = {path = "../..", develop = true}
```

**Review by bkeryan — CHANGES_REQUESTED** [review]

File: examples/sample_measurement/pyproject.toml (hunk 11 above)
Comment (dixonjoel) [reply examples/sample_measurement/pyproject.toml]: Sorry

Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/pyproject.toml:17]: Uncommented again so that the checked in code uses prerelease dependencies. I wasn't entirely sure if this was something we did locally or if we actually wanted this change in the repo.

File: examples/sample_measurement/pyproject.toml
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/pyproject.toml:1]: Updated to use 1.2.

**Review by bkeryan — APPROVED** [review]

File: examples/sample_measurement/pyproject.toml (hunk 11 above)
Comment (bkeryan) [reply examples/sample_measurement/pyproject.toml:17]: I want this change in the repo until we build a prerelease that has protobuf enum support, then we can comment it back out.

We need to figure out a better workflow for this. :(

File: examples/sample_measurement/pyproject.toml [hunk 14]
```diff
@@ -6,15 +6,21 @@ authors = ["National Instruments"]
...
+ni-measurementlink-service = {path = "../..", develop = true}
 
 [build-system]
 requires = ["poetry-core>=1.2.0"]
 build-backend = "poetry.core.masonry.api"
+
+[[tool.mypy.overrides]]
+module = [
+    "ni-measurementlink-service.*",
+]
```
Comment (bkeryan) [inline examples/sample_measurement/pyproject.toml:26]: This has no configuration keys so delete it. The module key identifies which module the other keys apply to.
https://mypy.readthedocs.io/en/stable/config_file.html#using-a-pyproject-toml-file

File: examples/sample_measurement/measurement.py (hunk 13 above)
Comment (bkeryan) [reply examples/sample_measurement/measurement.py:50]: Ok, I see what's going on.

The runtime type is `google.protobuf.internal.enum_type_wrapper.EnumTypeWrapper`, but the type used in `color_pb2.pyi` is `google.protobuf.internal.enum_type_wrapper._EnumTypeWrapper[_V]`:
```
class _ProtobufColor:
    ValueType = typing.NewType("ValueType", builtins.int)
    V: typing_extensions.TypeAlias = ValueType

class _ProtobufColorEnumTypeWrapper(google.protobuf.internal.enum_type_wrapper._EnumTypeWrapper[_ProtobufColor.ValueType], builtins.type):
    DESCRIPTOR: google.protobuf.descriptor.EnumDescriptor
    NONE: _ProtobufColor.ValueType  # 0
    PINK: _ProtobufColor.ValueType  # 1
    WHITE: _ProtobufColor.ValueType  # 2
    BLACK: _ProtobufColor.ValueType  # 3

class ProtobufColor(_ProtobufColor, metaclass=_ProtobufColorEnumTypeWrapper):
    """---------------------------------------------------------------------
    ---------------------------------------------------------------------
    """
```

In the type stubs, `EnumTypeWrapper` inherits from `_EnumTypeWrapper[int]`. `_EnumTypeWrapper[_V]` does not inherit from `EnumTypeWrapper`.
https://github.com/python/typeshed/blob/main/stubs/protobuf/google/protobuf/internal/enum_type_wrapper.pyi
```
_V = TypeVar("_V", bound=int)

# Expose a generic version so that those using mypy-protobuf
# can get autogenerated NewType wrapper around the int values
class _EnumTypeWrapper(Generic[_V]):
    DESCRIPTOR: EnumDescriptor
    def __init__(self, enum_type: EnumDescriptor) -> None: ...
    def Name(self, number: _V) -> str: ...
    def Value(self, name: str | bytes) -> _V: ...
    def keys(self) -> list[str]: ...
    def values(self) -> list[_V]: ...
    def items(self) -> list[tuple[str, _V]]: ...

class EnumTypeWrapper(_EnumTypeWrapper[int]): ...
```

This seems to work:
```
diff --git a/ni_measurementlink_service/measurement/service.py b/ni_measurementlink_service/measurement/service.py
index 3e70dc4..7ba372a 100644
--- a/ni_measurementlink_service/measurement/service.py
+++ b/ni_measurementlink_service/measurement/service.py
@@ -40,9 +40,9 @@ from ni_measurementlink_service.measurement.info import (
 from ni_measurementlink_service.session_management import PinMapContext

 if TYPE_CHECKING:
-    from google.protobuf.internal.enum_type_wrapper import EnumTypeWrapper
+    from google.protobuf.internal.enum_type_wrapper import EnumTypeWrapper, _EnumTypeWrapper, _V

-    SupportedEnumType = Union[Type[Enum], EnumTypeWrapper]
+    SupportedEnumType = Union[Type[Enum], EnumTypeWrapper, _EnumTypeWrapper[_V]]

```

Comment (bkeryan) [reply examples/sample_measurement/measurement.py:50]: Correction: I think this is sufficient. I don't think we need `EnumTypeWrapper` and `_V`.
```
SupportedEnumType = Union[Type[Enum], _EnumTypeWrapper]
```

File: examples/sample_measurement/pyproject.toml (hunk 14 above)
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/pyproject.toml]: Removed as suggested.

File: examples/sample_measurement/measurement.py (hunk 13 above)
Comment (WesleyTangNationalInstruments) [reply examples/sample_measurement/measurement.py]: Switched to `SupportedEnumType = Union[Type[Enum], _EnumTypeWrapper]` as suggested. Thanks for all the help on these issues.

File: examples/sample_measurement/pyproject.toml (hunk 12 above)
Comment (bkeryan) [reply examples/sample_measurement/pyproject.toml]: It should have ignored missing imports for `examples.*`
